Add reusable settings dialog to QDS Reports

SakshiS-harma opened the pull request:
This PR adds reusable components to Configuration dialog for different QDS Reports.

Configure dialog for Top Resource Consuming Queries:
![image](https://github.com/microsoft/azuredatastudio/assets/57200045/a3923999-d15b-4393-a3b5-29c667f0bf60)

SSMS equivalent:
![image](https://github.com/microsoft/azuredatastudio/assets/57200045/6beb0523-2a86-445a-9884-c5f969659ae9)

Configure dialog for Overall Resource Consumption:
![image](https://github.com/microsoft/azuredatastudio/assets/57200045/8321bbf5-b7e8-4093-8308-26247318e8da)

SSMS equivalent:
![image](https://github.com/microsoft/azuredatastudio/assets/57200045/12549475-3739-45b1-a5bf-884d66659187)

SakshiS-harma commented on extensions/query-store/src/common/utils.ts:
```diff
@@ -74,3 +74,8 @@ export function createSplitView(view: azdata.ModelView, firstComponent: azdata.C
 	return splitview;
 }
+
+export enum ConfigInfo {
```
This enum is just required to figure out distinction between the two reports to add Time Interval component appropriately. With a little duplication of code, it can be removed. Thoughts on which track to take?

Overall Resource Consumption:
![image](https://github.com/microsoft/azuredatastudio/assets/57200045/219d0825-b213-4b43-b3c8-1df33ffa87f5)

Top Resource Consuming Queries:
![image](https://github.com/microsoft/azuredatastudio/assets/57200045/d0ed92ad-a501-441a-a254-9b2688a93bc4)

kisantia commented:
@erinstellato-ms for any feedback on the configuration dialogs.

From looking at the screenshots, my initial thoughts are that it would look better if the input boxes are aligned. Also not sure if it's possible to break up the sections visually like in SSMS. The connection advanced properties dialog has something similar, but that's in core and not sure if that's possible in extensions. If it isn't, maybe having the section headers be bold, like in the Create project from db dialog, would help make the different sections more clear 
![image](https://github.com/microsoft/azuredatastudio/assets/31145923/44528979-9c85-4f7d-af7a-5c115d222793)

kisantia commented on extensions/query-store/src/reports/overallResourceConsumption.ts:
```diff
@@ -32,4 +34,10 @@ export class OverallResourceConsumption extends BaseQueryStoreReport {
 		return [durationContainer, executionCountContainer, cpuTimeContainer, logicalReadsContainer];
 	}
+
+	public override async configureButtonClick(configureDialog: ConfigureDialog): Promise<void> {
```
If this doesn't get called outside of the report, can this be protected instead of public?

kisantia commented on extensions/query-store/src/settings/configureDialog.ts:
```diff
@@ -0,0 +1,626 @@
+
+		this.dialog.okButton.label = constants.okButtonText;
+		this.dialog.okButton.enabled = false;
+		this.toDispose.push(this.dialog.okButton.onClick(async () => { }));		// TODO: Add return of settings change and functionality to enable OK Button
```
the ok, cancel, and apply onClicks should call dispose()

SakshiS-harma replied:
As far as I can understand, apply shouldn't be calling dispose- because the dialog is still open, so shouldn't yet be disposed.
As for, OK button click- adding dispose call (and the call exists for Cancel button).

kisantia commented on extensions/query-store/src/settings/configureDialog.ts:
```diff
@@ -0,0 +1,626 @@
+		this.dialog.cancelButton.label = constants.cancelButtonText;
+		this.toDispose.push(this.dialog.cancelButton.onClick(async () => await this.cancel()));
+
+		let applyButton = azdata.window.createButton(constants.applyButtonText);
```
what's the difference between the apply and ok button?

SakshiS-harma replied:
In SSMS, Apply button updates the UI with the changes made in the dialog (in the background), but keeps the config dialog open. Ok = Apply functionality + Close the dialog.
If it isn't necessary, I can remove it. I just wanted a parity between SSMS and ADS, hence this button :).

kisantia replied:
ah ok got it. A little confusing to a new person to me, but makes sense to have parity with SSMS.

SakshiS-harma replied:
On a different thought, I feel "Apply" along with "Ok" button will be confusing for ADS users. Even thought Ok + Cancel + Apply is a frequent siting in SSMS, but there is none in ADS. And Apply in ADS = Ok button. So for less confusion, I will remove the Apply button. And we can reintroduce it if it makes sense later.

SakshiS-harma replied:
Removed Apply button, since it doesn't exist in ADS (and can confuse ADS customers).

kisantia commented on extensions/query-store/src/settings/configureDialog.ts:
```diff
@@ -0,0 +1,626 @@
+				label: constants.rowCountLabel
+			}).component();
+
+		this.executionCountCheckBox.checked = true;
```
is the state of these saved somewhere? Does the dialog remember if one of these options is changed, applied, then the config dialog is reopened?

kisantia replied:
can you also add a todo comment in the code?

SakshiS-harma replied:
Will do. Also, added a work item to track it.

kisantia commented on extensions/query-store/src/settings/configureDialog.ts:
```diff
@@ -0,0 +1,626 @@
+			if (this.timeIntervalOptionsDropdown?.value === constants.customLabel) {
+				this.customTimeFromTextBox!.enabled = true;
+				await this.customTimeFromTextBox?.updateProperties({
+					value: '5/15/2023 11:58 AM'
```
add a todo comment to update these to not be hardcoded values

kisantia commented on extensions/query-store/src/settings/configureDialog.ts:
```diff
@@ -0,0 +1,626 @@
+			this.showChartComponent = this.createShowChartComponent();
+			this.timeIntervalComponent = this.createTimeIntervalComponent();
+			this.returnComponent = this.createReturnComponent();
+			this.filterComponent = this.createFilterComponent();
```
do these all need to be created here or can they be created only if the dialog needs it? Maybe the report can pass in a list of the components it needs when it creates the dialog?

SakshiS-harma replied:
Updated it to pass a list of components required and only those components are then created.

SakshiS-harma commented on extensions/query-store/src/common/uiConstants.ts:
```diff
@@ -0,0 +1,18 @@
+export namespace cssStyles {
+	export const text = { 'user-select': 'text', 'cursor': 'text' };
+	export const tableHeader = { ...text, 'text-align': 'left', 'border': 'none', 'font-size': '12px', 'font-weight': 'normal', 'color': '#666666' };
+	export const tableRow = { ...text, 'border-top': 'solid 1px #ccc', 'border-bottom': 'solid 1px #ccc', 'border-left': 'none', 'border-right': 'none', 'font-size': '12px' };
```
This is how it looks for now (not updated the alignment of the text boxes).
![image](https://github.com/microsoft/azuredatastudio/assets/57200045/92a9266c-4e0c-4bc4-b76d-a268ee7b6984)

![image](https://github.com/microsoft/azuredatastudio/assets/57200045/c03cd8fb-e4bd-4c1d-82ff-5b139aa942cc)

[And these UI constants aren't used for now, so will be removing them for next iteration].

SakshiS-harma replied:
I will add these to the main PR description once the alignment is fixed.

erinstellato-ms commented:
> @erinstellato-ms for any feedback on the configuration dialogs.
> 
> From looking at the screenshots, my initial thoughts are that it would look better if the input boxes are aligned. Also not sure if it's possible to break up the sections visually like in SSMS.

@kisantia @SakshiS-harma agreed on alignment and seeing if we can break the settings into sections in some way.  Let me know if you want me to create a mockup based on the existing screenshot, or if you want me to review after some changes.

SakshiS-harma commented:
> > @erinstellato-ms for any feedback on the configuration dialogs.
> > From looking at the screenshots, my initial thoughts are that it would look better if the input boxes are aligned. Also not sure if it's possible to break up the sections visually like in SSMS.
> 
> @kisantia @SakshiS-harma agreed on alignment and seeing if we can break the settings into sections in some way. Let me know if you want me to create a mockup based on the existing screenshot, or if you want me to review after some changes.

@erinstellato-ms I am working on one more change, and will look for your feedback after that. Hopefully that should look better than the current change :-) .

SakshiS-harma commented:
@erinstellato-ms I have updated the PR description with the latest changes. Let me know if the dialog makes sense and/or something should be modified.

kisantia commented:
Looks great with the updates! One small nit: would be nice to have a little (~10px?) space from the right side of the dropdown to the left side of the input boxes
![image](https://github.com/microsoft/azuredatastudio/assets/31145923/3ce324be-6df3-4fce-8e36-329cbbd6ef53)

kisantia commented on extensions/query-store/src/common/utils.ts:
```diff
@@ -75,7 +75,11 @@ export function createSplitView(view: azdata.ModelView, firstComponent: azdata.C
-export enum ConfigInfo {
-	overallResourceConfig,
-	topResourceConfig
+export enum ConfigComponentsInfo {
```
add doc comment that these are the available components that can be added to a config dialog

kisantia commented on extensions/query-store/src/settings/configureDialog.ts:
```diff
@@ -84,59 +81,80 @@ export class ConfigureDialog {
 
-	public async openDialog(): Promise<void> {
-		await this.initializeDialog();
+	public async openDialog(configComponentsInfo: ConfigComponentsInfo[]): Promise<void> {
```
add doc comment and mention if the order of the components in the array will be the order they're added to the dialog

kisantia commented on extensions/query-store/src/settings/configureDialog.ts:
```diff
@@ -365,29 +412,36 @@ export class ConfigureDialog {
-		this.timeIntervalOptionsOverallReport = [constants.lastHourLabel, constants.lastDayLabel, constants.last2DaysLabel, constants.lastWeekLabel, constants.lastMonthLabel, constants.last6MonthsLabel, constants.lastYearLabel, constants.customLabel];
+	private createTimeIntervalComponent(isOverallResourceReport: boolean = false): azdata.FormComponent {
+		let value;
+		if (isOverallResourceReport) {
```
we're starting with just these first two reports, but is this check still accurate when the other query store reports are added?

SakshiS-harma replied:
@erinstellato-ms , @kisantia Any thoughts on the above question? Basically there are different values that are populated for time intervals in different reports. Does it make sense to have all the reports present the same values or should I stick to the original ?

erinstellato-ms replied:
@SakshiS-harma I never noticed that the drop downs had different intervals.  I suspect it's because aggregating the data (for Overall Resource Consumption) for the last 5 minutes is not easy to calculate in some scenarios.  I'm not clear on why the drop downs are different for the third and fourth screenshots, which have the same time (Regressed Queries).

SakshiS-harma replied:
Got it. I think it would be best to keep the values as-is.
@kisantia as for your initial comment, for now the changes work. But when there will be more reports coming in, the logic here would need to be updated. I am adding a comment here for now to make sure to update it in future.

SakshiS-harma replied:
Actually, updated it in the next iteration here, so that less work is left for future 👍

kisantia commented on extensions/query-store/src/settings/configureDialog.ts:
```diff
@@ -453,53 +508,63 @@ export class ConfigureDialog {
 			.withLayout({ flexFlow: 'column' })
-			.withItems([timeIntervalFromRow, timeIntervalToRow])
+			.withItems([this.timeIntervalOptionsDropdown, timeIntervalFromRow, timeIntervalToRow])
+			.withProps({ ariaRole: 'timeIntervalGroup' })
```
I've only seen `ariaRole` set to `radiogroup` for a group of radio buttons and there's a list of well-known values for this property. Should this be `ariaLabel` instead?

SakshiS-harma replied:
Updated ariaRole to radiogroup (that was the original intent).

erinstellato-ms commented:
> Looks great with the updates! One small nit: would be nice to have a little (~10px?) space from the right side of the dropdown to the left side of the input boxes ![image](https://user-images.githubusercontent.com/31145923/263128210-3ce324be-6df3-4fce-8e36-329cbbd6ef53.png)

Agree that the changes look great!  I am actually opposite @kisantia - I was wondering about moving the date boxes more to the left, like this:

![image](https://github.com/microsoft/azuredatastudio/assets/98599126/793515a9-3c21-43bb-9c00-2a50c4be64f9)

SakshiS-harma commented:
@kisantia , @erinstellato-ms  Updated screenshots with latest changes. Please review.

erinstellato-ms commented:
@SakshiS-harma Screenshots look good!  To be consistent, should you update the time section for Top Resource Consumuing queries to Time Settings so it matches Overall Resource Consumption?  Also, I don't know if there is any way to shift the boxes (user entry points) to the left more, so there is less white space between the text and the boxes?  For Top Resource, we could abbreviate Minimum number of query plans to Min number of plans?

kisantia commented on extensions/query-store/src/common/utils.ts:
```diff
@@ -75,6 +75,10 @@ export function createSplitView(view: azdata.ModelView, firstComponent: azdata.C
 
+/**
+ * These are the available components that can be added to a config dialog.
+ * The sequence of the components here is the order that will be followed for addition to the report.
```
oh so this follows the order listed here, even if the array passed into openDialog() is a different order? I would've expected the order to be what's passed in the array. Ex:  passing in [timeIntervalComponent, chartComponent, filterComponent] would display it in that order, rather than [chartComponent, timeIntervalComponent, filterComponent]

SakshiS-harma replied:
That makes complete sense. I have updated the logic to follow the order provided in the list.

kisantia replied:
update comment here too with the change :)

SakshiS-harma commented:
> @SakshiS-harma Screenshots look good! To be consistent, should you update the time section for Top Resource Consumuing queries to Time Settings so it matches Overall Resource Consumption? Also, I don't know if there is any way to shift the boxes (user entry points) to the left more, so there is less white space between the text and the boxes? For Top Resource, we could abbreviate Minimum number of query plans to Min number of plans?

Updated according to your comments here. I have updated the screenshots as well. Let me know if the white space should be reduced more.
